engine/src/scheme.rs

```diff
@@ -221,29 +314,121 @@ impl<'s> Function<'s> {
     pub(crate) fn as_definition(&self) -> &'s dyn FunctionDefinition {
         &*self.scheme.inner.functions[self.index].1
     }
+
+    /// Converts to an owned [`Function`].
+    #[inline]
+    pub fn to_owned(&self) -> Function {
+        Function {
+            scheme: self.scheme.clone(),
+            index: self.index,
+        }
+    }
+
+    /// Reborrows the function relatively to the specified [`Scheme`] reference.
+    ///
+    /// Useful when you have a [`FunctionRef`] borrowed from an owned [`Function`]
+    /// but you need to extend/change it's lifetime.
+    ///
+    /// Panics if the function doesn't belong to the specified scheme.
+    #[inline]
+    pub fn reborrow(self, scheme: &Scheme) -> FunctionRef<'_> {
+        assert!(self.scheme == scheme);
+
+        FunctionRef {
+            scheme,
+            index: self.index,
+        }
+    }
+}
+
+impl PartialEq<Function> for FunctionRef<'_> {
+    #[inline]
+    fn eq(&self, other: &Function) -> bool {
+        self.eq(&other.as_ref())
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Hash)]
+/// A structure to represent a function inside a [`Scheme`](struct@Scheme).
+pub struct Function {
+    scheme: Scheme,
+    index: usize,
+}
+
+impl Serialize for Function {
+    fn serialize<S: Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
+        self.name().serialize(ser)
+    }
+}
+
+impl Debug for Function {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+impl Function {
+    /// Returns the function's name as recorded in the [`Scheme`](struct@Scheme).
+    #[inline]
+    pub fn name(&self) -> &str {
+        &self.scheme.inner.functions[self.index].0
+    }
+
+    /// Get the function's index in the [`Scheme`](struct@Scheme) identifier's list.
+    #[inline]
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// Returns the [`Scheme`](struct@Scheme) to which this function belongs to.
+    #[inline]
+    pub fn scheme(&self) -> &Scheme {
+        &self.scheme
+    }
+
+    #[inline]
+    pub(crate) fn as_definition(&self) -> &dyn FunctionDefinition {
+        &*self.scheme.inner.functions[self.index].1
+    }
+
+    /// Converts to a borrowed [`Function`].
+    #[inline]
+    pub fn as_ref(&self) -> FunctionRef<'_> {
+        FunctionRef {
+            scheme: &self.scheme,
+            index: self.index,
+        }
+    }
+}
+
+impl PartialEq<FunctionRef<'_>> for Function {
+    #[inline]
+    fn eq(&self, other: &FunctionRef<'_>) -> bool {
+        self.as_ref().eq(other)
+    }
 }
 
 /// An enum to represent an entry inside a [`Scheme`](struct@Scheme).
 /// It can be either a [`Field`](struct@Field) or a [`Function`](struct@Function).
 #[derive(Debug)]
 pub enum Identifier<'s> {
```

Comment:
Should this be called `IdentifierRef` now?

Reply:
Its not exposed anymore since https://github.com/cloudflare/wirefilter/pull/122 so I am not sure it matters much.
I actually wonder if we should just get rid of that type entirely.